### PR TITLE
Add IRouteHandler, RouteBase, and IRoutingFeature

### PR DESCRIPTION
**Pull request by rynowak**
Adds IRouterHandler, an abstraction for endpoints in the routing system
that can't chain (example: delegates). The idea is that some kinds of
routes aren't really friendly to chaining. If you don't support chaining,
then accept IRouteHandler and work with that rather than IRouter.

There's one implementation of IRouteHandler, RouteHandler. It implements
both IRouter and IRouteHandler.

Adds RouteBase as a base class for routes based on our template syntax and
defaults/constraints/data-tokens. Updated a lot of signatures to be
get/set virtual and mutable to facilitate or bigger variety of usage
scenarios.

Renamed TemplateRoute to just Route, now inherits from RouteBase.

Adds IRoutingFeature for middleware scenarios where you don't have access
to the route context.

Also adds some basic extension methods for accessing route values.

### CODE REVIEW

File: src/Microsoft.AspNet.Routing/Internal/TaskCache.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNet.Routing.Internal
+{
+    public static class TaskCache
+    {
+        public static readonly Task CompletedTask = Task.FromResult(0);
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Routing/Internal/TaskCache.cs]: Isn't there a Task.Completed in CoreCLR?

Comment (rynowak) [inline src/Microsoft.AspNet.Routing/Internal/TaskCache.cs]: Yeah. I can put in an ifdef here.

File: src/Microsoft.AspNet.Routing.Abstractions/IRouteHandler.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNet.Http;
+
+namespace Microsoft.AspNet.Routing
+{
+    public interface IRouteHandler
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Routing.Abstractions/IRouteHandler.cs]: Docs

File: src/Microsoft.AspNet.Routing/Route.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+            _target = target;
+        }
+
+        public string RouteTemplate => ParsedTemplate.TemplateText;
+
+        protected override async Task OnRouteMatched(RouteContext context)
+        {
+            context.RouteData.Routers.Add(_target);
+            await _target.RouteAsync(context);
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Routing/Route.cs]: `return _target.RouteAsync(context);`

File: src/Microsoft.AspNet.Routing/Tree/TreeRouteLinkGenerationEntry.cs [hunk 4]
```diff
@@ -20,12 +20,12 @@ public class TreeRouteLinkGenerationEntry
         /// <summary>
         /// The route constraints.
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Routing/Tree/TreeRouteLinkGenerationEntry.cs]: Gets or sets

File: test/Microsoft.AspNet.Routing.Tests/RouteTest.cs [hunk 5]
```diff
@@ -730,25 +472,12 @@ public void CreateTemplate_InlineConstraint_Regex_Malformed()
...
-            var context = CreateRouteContext("/Home");
-
-            // Act
-            await route.RouteAsync(context);
+            Assert.Null(context.Handler);
 
-            // Assert
-            Assert.False(context.IsHandled);
-            Assert.Empty(context.RouteData.Routers);
+            Assert.Single(context.RouteData.Values);
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Routing.Tests/RouteTest.cs]:
```C#
var value = Assert.Single(context.RouteData.Values);
Assert.Equal("controller", value.Key);
Assert.Equal("Home", Assert.IsType<string>(value.Value));
```

**Comment by pranavkm** [comment]
Can you add docs to all the new API in abstractions?

:shipit: once that is addressed.

**Pull request closed without merging**